sdks/python/apache_beam/runners/interactive/pipeline_instrument.py

```diff
@@ -0,0 +1,470 @@
+    not exist in current cache.
+
+    Modifies:
+      self._pipeline
+    """
+    if pcoll.pipeline is not self._pipeline:
+      return
+    key = self.cache_key(pcoll)
+    if not self._cache_manager.exists('full', key):
+      logging.debug('%s write cache: %s', pcoll.producer, key)
+      _ = pcoll | '{}{}'.format(WRITE_CACHE, key) >> cache.WriteCache(
+          self._cache_manager, key)
+
+  def _read_cache(self, pcoll):
+    """Reads a cached pvalue.
+
+    A noop will cause the pipeline to execute the transform as
+    it is and cache nothing from this transform for next run.
+
+    Modifies:
+      self._pipeline
+    """
+    if pcoll.pipeline is not self._pipeline:
+      return
+    key = self.cache_key(pcoll)
+    if self._cache_manager.exists('full', key):
+      if key not in self._cached_pcoll_read:
+        logging.debug('read cache: %s', key)
+        # Mutates the pipeline with cache read transform attached
+        # to root of the pipeline.
+        pcoll_from_cache = (
+            self._pipeline
+            | '{}{}'.format(READ_CACHE, key) >> cache.ReadCache(
+                self._cache_manager, key))
+        self._cached_pcoll_read[key] = pcoll_from_cache
+    # else: NOOP when cache doesn't exist.
+
+  def _replace_with_cached_inputs(self):
+    """Replace PCollection inputs in the pipeline with cache if possible.
+
+    For any input PCollection, find out whether there is valid cache. If so,
+    replace the input of the AppliedPTransform with output of the
+    AppliedPtransform that sources pvalue from the cache. If there is no valid
+    cache, noop.
+    """
+
+    class ReadCacheWireVisitor(PipelineVisitor):
+      """Visitor wires cache read as inputs to replace corresponding original
+      input PCollections in pipeline.
+      """
+
+      def __init__(self, pin):
+        """Initializes with a PipelineInstrument."""
+        self._pin = pin
+
+      def enter_composite_transform(self, transform_node):
+        self.visit_transform(transform_node)
+
+      def visit_transform(self, transform_node):
+        if transform_node.inputs:
+          input_list = list(transform_node.inputs)
+          for i in range(len(input_list)):
+            key = self._pin.cache_key(input_list[i])
+            if key in self._pin._cached_pcoll_read:
+              input_list[i] = self._pin._cached_pcoll_read[key]
+          transform_node.inputs = tuple(input_list)
+
+    v = ReadCacheWireVisitor(self)
+    self._pipeline.visit(v)
+
+  def _cacheable_inputs(self, transform):
+    inputs = set()
+    for in_pcoll in transform.inputs:
+      if self._cacheable_key(in_pcoll) in self.cacheables():
+        inputs.add(in_pcoll)
+    return inputs
+
+  def _cacheable_key(self, pcoll):
+    """Gets the key a cacheable PCollection is tracked within the instrument."""
+    return cacheable_key(pcoll, self.pcolls_to_pcoll_id(),
+                         self._pcoll_version_map)
+
+  def cache_key(self, pcoll):
```

Comment:
What is the difference between cacheable_key and cache_key, they both seemingly return a key.

Reply:
A cacheable_key is the key to a cacheable.
A cacheable means that the PCollection described by it could be cached. But whether a cache exists (a pipeline has been executed before, a PCollection has been re-evaluated), the pipeline_instrument doesn't know.

A cache_key is the key to a cache.
When given a cache_key, if caching mechanism cannot find a cache with that key. It means that the cache has not been created. And this cache_key will eventually lead you to some cache as pipeline executions occur in current notebook environment, until the pipeline_instrument stops generating this cache_key but other new keys (caused by changes in the user defined pipeline in notebook). Old cache is invalidated in this way.

Because there are 2 context:
1. In notebook, there is an `apache_beam.Pipeline` instance where user has a handle.
2. In runner and this pipeline_instrument, there is a copied `apache_beam.Pipeline` instance where the runner and pipeline_instrument has a handle and potentially mutates.
Context 1 isolates the scope of caching. When the `pipeline` object is gone in context 1, all the caching will be invalidated. 
Example scenarios to refresh context 1:
1. Restarts the IPython kernel;
2. Assign a different instance to `pipeline` in notebook.

Within the scope of context 2, pipeline_instrument accesses context 1 to figure out the relationship between PCollections in copied `pipeline` (in context 2) and original `pipeline` (in context 1) to decide where the cache should be written to or read from.

cache_key functions anywhere (pipeline execution, PCollection visualization or simply read a cache file) within context 1.
cacheable_key points to cacheable linking context 1 and context 2 within a single `PipelineInstrument` instance. The instance is only valid during `pipeline.run()`.
Whenever the `pipeline` changes in context 1, this `PipelineInstrument` should also be re-evaluated (within InteractiveRunner).

Reply:
I understand. But this is very confusing. I do suggest simplification of terms, concepts and improvement of docs. Some of your explanations could be comments here.

I would not block the PR but please make an overall pass to update the readability. Ideally it should be possible for another engineer to come, learn and be able to contribute these code paths as easily as possible.